src/main/java/org/spongepowered/api/plugin/Plugin.java

```diff
@@ -24,19 +24,21 @@
  */
 package org.spongepowered.api.plugin;
 
-import static java.lang.annotation.ElementType.TYPE;
-import static java.lang.annotation.RetentionPolicy.RUNTIME;
-
+import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
+import java.util.regex.Pattern;
 
 /**
  * An annotation used to describe and mark a Sponge plugin.
  */
-@Target(TYPE)
-@Retention(RUNTIME)
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
 public @interface Plugin {
 
+    Pattern ID_PATTERN = Pattern.compile("[a-z0-9-_]+");
```

Comment:
That sounds good. I'd still like one (not both) of these options so we have a basic level of verification and making sure plugins are usually fully qualified (which is the recommended way):

- Add `group` separately to the `@Plugin` annotation. This would make the group required, therefore it's probably not a good option.

- Add warning in plugin annotation processor if the plugin ID is not qualified. Require qualified plugin IDs on Ore. Considering there are plans to add a builtin Maven repository for plugins hosted on Ore this sounds reasonable because we ensure public plugins use the proper qualified IDs and there is also always a proper group and ID available for the Maven artifact.

  - This also follows the design of file metadata which is recommended but not required for plugins the way I currently designed it. Still, the annotation processor and the implementations will print a warning when there is no file metadata defined, and it's generally supposed to be required for plugins on Ore as well.

I don't think we should recommend anything else than qualified IDs because the plugin IDs are used internally in many ways, e.g. for commands and anything not related to the plugin name would just confuse all users. Speaking of commands, it would be probably reasonable to provide the commands as non-qualified plugin ID (as well as qualified?) so you can also type `/myplugin:test` instead of `/com.mumfrey.plugins.myplugin:test`.